pyrsa/rdm/compare.py

```diff
@@ -0,0 +1,178 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Fri Dec  6 13:01:12 2019
+
+@author: heiko
+"""
+import numpy as np
+import scipy.stats
+
+
+def compare_cosine(rdm1, rdm2):
+    """
+    calculates the cosine distance between two RDMs objects
+
+        Args:
+            rdm1 (pyrsa.rdm.RDMs):
+                first set of RDMs
+            rdm2 (pyrsa.rdm.RDMs):
+                second set of RDMs
+        Returns:
+            dist (float):
+                cosine distance between the two RDMs
+    """
+    vector1, vector2 = _parse_input_rdms(rdm1, rdm2)
+    dist = _average_all_combinations(vector1, vector2, _cosine)
+    return 1 - dist
+
+
+def compare_correlation(rdm1, rdm2):
+    """
+    calculates the correlation distance between two RDMs objects
+
+        Args:
+            rdm1 (pyrsa.rdm.RDMs):
+                first set of RDMs
+            rdm2 (pyrsa.rdm.RDMs):
+                second set of RDMs
+        Returns:
+            dist (float):
+                correlation distance between the two RDMs
+    """
+    vector1, vector2 = _parse_input_rdms(rdm1, rdm2)
+    vector1 = vector1 - np.mean(vector1, 1, keepdims=True)
+    vector2 = vector2 - np.mean(vector2, 1, keepdims=True)
+    dist = _average_all_combinations(vector1, vector2, _cosine)
+    return 1 - dist
+
+
+def compare_rank_corr(rdm1, rdm2):
+    """
+    calculates the correlation distance between two RDMs objects
+
+        Args:
+            rdm1 (pyrsa.rdm.RDMs):
+                first set of RDMs
+            rdm2 (pyrsa.rdm.RDMs):
+                second set of RDMs
+        Returns:
+            dist (float):
+                rank correlation distance between the two RDMs
+    """
+    vector1, vector2 = _parse_input_rdms(rdm1, rdm2)
+    dist = _average_all_combinations(vector1, vector2, _spearman_r)
+    return 1 - dist
+
+
+def compare_kendall_tau(rdm1, rdm2):
+    """
+    calculates the Kendall-tau based distance between two RDMs objects
+
+        Args:
+            rdm1 (pyrsa.rdm.RDMs):
+                first set of RDMs
+            rdm2 (pyrsa.rdm.RDMs):
+                second set of RDMs
+        Returns:
+            dist (float):
+                kendall-tau based distance between the two RDMs
+    """
+    vector1, vector2 = _parse_input_rdms(rdm1, rdm2)
+    dist = _average_all_combinations(vector1, vector2, _kendall_tau)
+    return 1 - dist
+
+
+def _average_all_combinations(vectors1, vectors2, func):
+    """
+    runs a function func on all combinations of v1 in vectors1
+    and v2 in vectors2 and averages the results
+
+        Args:
+            vectors1 (numpy.ndarray):
+                first set of values
+            vectors1 (numpy.ndarray):
+                second set of values
+            func (function):
+                function to be applied, should take two input vectors
+                and return one scalar
+        Returns:
+            value (float):
+                average function result over all pairs
+    """
+    sum_val = 0
```

Comment:
well, essentially only that I did not find a function in numpy which does what I want. I think yours would assume that rdm1 and rdm2 have equally many entries and then only compare the matching ones? I also thought that there should be some kind of outer product implementation which works like this function, but I was unable to locate it. If anyone knows, please tell me and I will replace this.

Reply:
Ah yes I assumed they would be the same size. I didn't realize you could compare RDMs of different sizes.

Reply:
well, different numbers of RDMs not different sizes of RDMs in this case

Reply:
Anyway: If anyone out there knows how to do this with numpy/scipy, please tell me, otherwise I'll assume this is resolved once another review comes in.

Reply:
But if it's about missing data don't you want to just output NaNs where appropriate?

Reply:
Hmmm yes, but that is not the case here? The aim here is: I have two sets of RDMs: For example from different subjects and from different instances of a model. I cannot simply average them, because they may have different scaling, which I want to ignore. To solve this, I correlate each RDM from group 1 with each RDM from group 2 and average the results. 

This is more akin to a generalized outer product + then average than to applying along a dimension & I did in fact not care about missing values yet.

Reply:
Ah..ok I think you lost me :) but code wise :+1:

Reply:
Hi @HeikoSchuett i'm not sure I follow what you are trying to do. So if you have 2 sets of RDMs. say one set where you have 3x2556 (3 regions of interest, 3 subjects, whatever..) and you have 8 models 8x2556, you want to get back a 3x8 of correlations?

Reply:
Would a function like this one help?
```python
def corr_rdms(X, Y):
    """ corr_rdms correlate two stacks of RDMs.
        
    Args:
        X [array]: e.g. a stack of fmri ROI RDMs (n_rois x n_pairs)
        Y [array]: e.g. a stack of model RDMs (n_layers x n_pairs)

    Returns:
        [array]: correlations between X and Y, of shape dim0 of X by dim0 of Y
    """
    X = X - X.mean(axis=1, keepdims=True)
    X /= np.sqrt(np.einsum('ij,ij->i', X, X))[:, None]
    Y = Y - Y.mean(axis=1, keepdims=True)
    Y /= np.sqrt(np.einsum('ij,ij->i', Y, Y))[:, None]

    return np.einsum('ik,jk', X, Y)
```

Reply:
Hi @iancharest, 
here I immediately averaged to get a single value for the comparison of two rdms objects. Your comments got me thinking that your version and averaging if we need to elsewhere might be a better idea.  

I definitely like your corr_rdms function. I'll go and switch to that way of calculating for cosine and correlation distance.

Reply:
Hi @iancharest,
changes are now implemented. Does his seem right to you?